File: DataQuality/DataQualityTools/src/DQTLumiMonAlg.cxx
/*
  Copyright (C) 2002-2021 CERN for the benefit of the ATLAS collaboration
*/

#include "DataQualityTools/DQTLumiMonAlg.h"

DQTLumiMonAlg::DQTLumiMonAlg(const std::string& name, ISvcLocator* pSvcLocator)
: AthMonitorAlgorithm(name,pSvcLocator)
{}

DQTLumiMonAlg::~DQTLumiMonAlg() {}

StatusCode DQTLumiMonAlg::initialize(){
    // need to do this first, else we don't get environment() set up properly!
    ATH_CHECK( AthMonitorAlgorithm::initialize() );
    ATH_CHECK( m_VertexContainerKey.initialize() );
    ATH_CHECK( m_PixelClustersKey.initialize( environment() != Environment_t::AOD ) );
    ATH_CHECK( m_PixelIDKey.initialize() );
    return StatusCode::SUCCESS;
}

StatusCode DQTLumiMonAlg::fillHistograms(const EventContext& ctx) const {
    using namespace Monitored;
    auto group = getGroup("default");

    RH<xAOD::EventInfo> eventInfo{GetEventInfo(ctx)};
    if ( !eventInfo.isValid() ) {
        ATH_MSG_WARNING("Could not retrieve EventInfo.");
        return StatusCode::FAILURE;
    }

    auto lumiBlock = Scalar<int>("LB", eventInfo->lumiBlock());

    auto avgLumi = Scalar("avgLumi", lbAverageLuminosity(ctx));
    auto avgIntPerXing = Scalar("avgIntPerXing", lbAverageInteractionsPerCrossing(ctx));
    float avgMuInverse = avgIntPerXing>0 ? 1./avgIntPerXing : 0.;
    auto lumiPerBCID = Scalar("lumiPerBCID", lbLuminosityPerBCID(ctx));
    auto intPerXing = Scalar("intPerXing", lbInteractionsPerCrossing(ctx));
    auto duration = Scalar("duration", lbDuration(ctx));
    auto avgLiveFrac = Scalar("avgLiveFrac", lbAverageLivefraction(ctx));
    auto liveFracPerBCID = Scalar("liveFracPerBCID", livefractionPerBCID(ctx));
    auto lumiWeight = Scalar("lumiWeight", lbLumiWeight(ctx));

    RH<xAOD::VertexContainer> vertices(m_VertexContainerKey,ctx);
    auto nLooseVtx = Scalar<int>("nLooseVtx",0);
    auto nTightVtx = Scalar<int>("nTightVtx",0);
    if ( vertices.isValid() ) {
        for ( const auto vertex : *vertices ) {
            if (!vertex || !vertex->nTrackParticles()) continue;
            nLooseVtx++;

            int nGoodTracks = 0;
            for (size_t i = 0; i < vertex->nTrackParticles(); ++i) {
                if (vertex->trackWeight(i) >= m_tightTrackWeight) { ++nGoodTracks; }
            }
            if ( nGoodTracks>=m_tightNTracks ) {
                nTightVtx++;
            }
        }
    } else {
        ATH_MSG_WARNING("Could not retrieve Vertex Container.");
    }
    auto nLooseVtxPerAvgMu = Scalar<float>("nLooseVtxPerAvgMu",nLooseVtx*avgMuInverse);
    auto nTightVtxPerAvgMu = Scalar<float>("nTightVtxPerAvgMu",nTightVtx*avgMuInverse);

    fill(group,lumiBlock,avgLumi,avgIntPerXing,lumiPerBCID,intPerXing,duration,
         avgLiveFrac,liveFracPerBCID,lumiWeight,nLooseVtx,nTightVtx,nLooseVtxPerAvgMu,
         nTightVtxPerAvgMu);

    if ( m_environment!=Environment_t::AOD ) {
        RH<InDet::PixelClusterContainer> pixelClusters(m_PixelClustersKey,ctx);
        RH<PixelID> pixelID(m_PixelIDKey,ctx);
        if ( !pixelClusters.isValid() ) {
            ATH_MSG_WARNING("Could not retrieve Pixel Clusters.");
        } else if ( !pixelID.isValid() ) {
            ATH_MSG_WARNING("Could not retrieve Pixel ID.");
        } else {
            auto nClustersAll = Scalar<int>("nClustersAll",0);
            auto nClustersECA = Scalar<int>("nClustersECA",0);
            auto nClustersECC = Scalar<int>("nClustersECC",0);
            auto nClustersB0 = Scalar<int>("nClustersB0",0);
            auto nClustersB1 = Scalar<int>("nClustersB1",0);
            auto nClustersB2 = Scalar<int>("nClustersB2",0);
            for ( auto cluster : *pixelClusters ) {
                if (!cluster)
                    continue;
                for ( auto pixel : *cluster ) {
                    Identifier clusterID = pixel->identify();
                    nClustersAll++;
                    int ec = pixelID->barrel_ec(clusterID);
                    if (ec==2) nClustersECA++;
                    else if (ec==-2) nClustersECC++;
                    else if (ec==0) {
                        int ld = pixelID->layer_disk(clusterID);
                        if (ld==0) nClustersB0++;
                        else if(ld==1) nClustersB1++;
                        else if(ld==2) nClustersB2++;
                    }
                }
                ATH_MSG_DEBUG("nClustersAll is " << nClustersAll);
            }
            auto nClustersAllPerAvgMu = Scalar<int>("nClustersAllPerAvg",nClustersAll*avgMuInverse);
            auto nClustersECAPerAvgMu = Scalar<int>("nClustersECAPerAvg",nClustersECA*avgMuInverse);
            auto nClustersECCPerAvgMu = Scalar<int>("nClustersECCPerAvg",nClustersECC*avgMuInverse);
            auto nClustersB0PerAvgMu = Scalar<int>("nClustersB0PerAvg",nClustersB0*avgMuInverse);
            auto nClustersB1PerAvgMu = Scalar<int>("nClustersB1PerAvg",nClustersB1*avgMuInverse);
            auto nClustersB2PerAvgMu = Scalar<int>("nClustersB2PerAvg",nClustersB2*avgMuInverse);

            fill("pixel",lumiBlock,
                 nClustersAll,nClustersECA,nClustersECC,
                 nClustersB0,nClustersB1,nClustersB2,
                 nClustersAllPerAvgMu,nClustersECAPerAvgMu,nClustersECCPerAvgMu,
                 nClustersB0PerAvgMu,nClustersB1PerAvgMu,nClustersB2PerAvgMu);
        }
    }

    return StatusCode::SUCCESS;
}

File: DataQuality/DataQualityTools/python/DQTLumiMonAlg.py
#
#  Copyright (C) 2002-2019 CERN for the benefit of the ATLAS collaboration
#

def DQTLumiMonAlgConfig(flags, isOld=False):
    if isOld: # replace with the old version of the tool
        from AthenaMonitoring import AthMonitorCfgHelperOld as AthMonitorCfgHelper
        from .DataQualityToolsConf import DQTLumiMonAlg
    else:
        from AthenaMonitoring import AthMonitorCfgHelper
        from AthenaConfiguration.ComponentFactory import CompFactory
        DQTLumiMonAlg = CompFactory.DQTLumiMonAlg


    helper = AthMonitorCfgHelper(flags, 'DQTLumiMonAlgCfg')
    # Three instances of the algorithm. One using any trigger, another using only muon
    # triggers, and the final using only electron triggers.
    DQTLumiMonAlgConfigByTriggerChain(helper, DQTLumiMonAlg, flags, isOld)
    if (flags.useTrigger if isOld else flags.DQ.useTrigger):
        DQTLumiMonAlgConfigByTriggerChain(helper, DQTLumiMonAlg, flags, isOld,
                                        'CATEGORY_monitoring_muonIso','EF_muX')
        DQTLumiMonAlgConfigByTriggerChain(helper, DQTLumiMonAlg, flags, isOld,
                                        'CATEGORY_primary_single_ele','EF_eX')
    if isOld:
        return helper.result()
    else:
        result = helper.result()
        from AtlasGeoModel.AtlasGeoModelConfig import AtlasGeometryCfg
        result.merge(AtlasGeometryCfg(flags))
        return result

def DQTLumiMonAlgConfigByTriggerChain(helper, algConfObj, flags, isOld, triggerChain='', triggerPath=''):
    monAlg = helper.addAlgorithm(algConfObj, 'DQTLumiMonAlg'+triggerPath)

    if triggerChain:
        monAlg.TriggerChain = triggerChain
        if not triggerPath:
            triggerPath = triggerChain
    else:
        triggerPath = 'AnyTrigger'

    groupPath = 'GLOBAL/Luminosity/'+triggerPath
    group = helper.addGroup(monAlg, 'default', groupPath, 'run')
    pixelgroup = helper.addGroup(monAlg, 'pixel', groupPath, 'run')
    nLooseT = 'Number of Loose Vertices'
    nTightT = 'Number of Tight Vertices'
    nPixClusterT = 'Number of Pixel Clusters'

    # Vertex parameters. A vertex
    monAlg.TightTrackWeight = 0.01
    monAlg.TightNTracks = 4

    if isOld:
        from RecExConfig.AutoConfiguration import GetLBNumber
        lbnum = GetLBNumber()
        lbdict = { 'xmin': lbnum-0.5 if lbnum>0 else 0.5,
                   'xmax': lbnum+0.5 if lbnum>0 else 1.5 }
    else:
        lbdict = { 'xmin': min(flags.Input.LumiBlockNumber)-0.5 if flags.Input.LumiBlockNumber else 0.5,
                   'xmax': max(flags.Input.LumiBlockNumber)+0.5 if flags.Input.LumiBlockNumber else 1.5 }
    lbdict['xbins'] = int(lbdict['xmax']-lbdict['xmin'])

    # Raw plots of lumi variables
    group.defineHistogram('avgIntPerXing', type='TH1F', xbins=100, xmin=0, xmax=100,
                          title='Average number of interactions per bunch Xing')
    group.defineHistogram('intPerXing', type='TH1F', xbins=100, xmin=0, xmax=100,
                          title='Actual number of interactions per bunch Xing')

    # Plots of luminosity variables as a function of lb number
    group.defineHistogram('LB,avgLumi', type='TProfile',
                          title='Average Lumi vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,avgIntPerXing', type='TProfile',
                          title='Number of interactions per event;LB;<#mu>_{LB}',
                          opt='kAddBinsDynamically', merge='merge', **lbdict)
    group.defineHistogram('LB,lumiPerBCID', type='TProfile',
                          title='Instantaneous Luminosity vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,intPerXing', type='TProfile',
                          title='Instaneous interactions vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,duration', type='TProfile',
                          title='Lumi Block time in sec vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,avgLiveFrac', type='TProfile',
                          title='Average live fraction lumi over all BCIDs vs LB',
                          opt='kAddBinsDynamically', merge='merge', **lbdict)
    group.defineHistogram('LB,liveFracPerBCID', type='TProfile',
                          title='Current BCID lumi vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,lumiWeight', type='TProfile',
                          title='Current BCID lumi vs LB', opt='kAddBinsDynamically',
                          merge='merge', **lbdict)

    # Vertex counts vs. lb number (loose and tight)
    group.defineHistogram('LB,nLooseVtx', type='TProfile',
                          title=nLooseT+' per event;LB;<NlooseVtx/event>_{LB}',
                          opt='kAddBinsDynamically', merge='merge', **lbdict)
    group.defineHistogram('LB,nTightVtx', type='TProfile',
                          title=nTightT+' per event;LB;<NtightVtx/event>_{LB}',
                          opt='kAddBinsDynamically', merge='merge', **lbdict)

    # Vertex counts per avg mu vs. lb number. (Same as above, just weighted.)
    group.defineHistogram('LB,nLooseVtxPerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                          title=nLooseT+' per event per Mu;LB;<NlooseVtx/event/#mu>_{LB}',
                          merge='merge', **lbdict)
    group.defineHistogram('LB,nTightVtxPerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                          title=nTightT+' per event per Mu;LB;<NtightVtx/event/#mu>_{LB}',
                          merge='merge', **lbdict)

    # Vertex counts vs. avg mu.
    group.defineHistogram('avgIntPerXing,nLooseVtx', type='TProfile', xbins=250, xmin=0, xmax=100,
                          title=nLooseT+' per event;#mu;NlooseVtx/event')
    group.defineHistogram('avgIntPerXing,nTightVtx', type='TProfile', xbins=250, xmin=0, xmax=100,
                          title=nTightT+' per event;#mu;NtightVtx/event')

    # Pixel clusters vs lb number. All, and then eparately for each endcap and barrel layer.
    pixelgroup.defineHistogram('LB,nClustersAll', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', all;LB;<NclustersAll/event>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersECA', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', endcap A;LB;<NlustersECA/event>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersECC', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', endcap C;LB;<NclustersECC/event>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB0', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', barrel layer 0;LB;<NclustersB0/event>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB1', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', barrel layer 1;LB;<NclustersB1/event>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB2', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+', barrel layer 2;LB;<NclustersB2/event>_{LB}',
                               merge='merge', **lbdict)

    # Pixel clusters per avg mu vs. lb number. (Same as above, just weighted.)
    pixelgroup.defineHistogram('LB,nClustersAllPerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, all;LB;<NclustersAll/event/#mu>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersECAPerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, endcap A;LB;<NlustersECA/event/#mu>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersECCPerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, endcap C;LB;<NclustersECC/event/#mu>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB0PerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, barrel layer 0;LB;<NclustersB0/event/#mu>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB1PerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, barrel layer 1;LB;<NclustersB1/event/#mu>_{LB}',
                               merge='merge', **lbdict)
    pixelgroup.defineHistogram('LB,nClustersB2PerAvgMu', type='TProfile', opt='kAddBinsDynamically',
                               title=nPixClusterT+' per Mu, barrel layer 2;LB;<NclustersB2/event/#mu>_{LB}',
                               merge='merge', **lbdict)

    # Pixel clusters vs. avg mu. All, then separately for each endcap and barrel layer.
    pixelgroup.defineHistogram('avgIntPerXing,nClustersAll', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', all;#mu;NclustersAll/event')
    pixelgroup.defineHistogram('avgIntPerXing,nClustersECA', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', endcap A;#mu;NclustersECA/event')
    pixelgroup.defineHistogram('avgIntPerXing,nClustersECC', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', endcap C;#mu;NclustersECC/event')
    pixelgroup.defineHistogram('avgIntPerXing,nClustersB0', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', barrel layer 0;#mu;NclustersB0/event')
    pixelgroup.defineHistogram('avgIntPerXing,nClustersB1', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', barrel layer 1;#mu;NclustersB1/event')
    pixelgroup.defineHistogram('avgIntPerXing,nClustersB2', type='TProfile',
                               xbins=250, xmin=0, xmax=100,
                               title=nPixClusterT+', barrel layer 2;#mu;NclustersB2/event')
